lldb/source/Target/SwiftLanguageRuntimeNames.cpp

```diff
@@ -245,16 +242,33 @@ class ThreadPlanStepInAsync : public ThreadPlan {
     s->PutCString("ThreadPlanStepInAsync");
   }
 
-  // Composite thread plans never directly explain a stop.
-  bool DoPlanExplainsStop(Event *event) override { return false; }
+  bool DoPlanExplainsStop(Event *event) override {
+    if (!HasTID())
+      return false;
+
+    if (!m_async_breakpoint_sp)
+      return false;
+
+    return GetBreakpointAsyncContext() == m_initial_async_ctx;
+  }
+
+  bool ShouldStop(Event *event) override {
+    if (!m_async_breakpoint_sp)
```

Comment:
It looks like if GetBreakpointAsyncContext matches this context the plan is complete.  So if GetBreakpointAsyncContext is expensive, you can do the SetPlanComplete in DoPlanExplainsStop, and then have ShouldStop check IsPlanComplete and use that to decide what to return.  That way you only have to do the work once.

There's one slightly non-obvious bit to that which is that is it possible that your plan wasn't the one that explained the stop.  Rather, an earlier plan explained the stop, but then that planwas finished, and and got popped off the stack.  We don't do another round of ExplainsStop at that point (*) but just ask the next plan on the stack if it ShouldStop.

In that case, your ShouldStop will get called, without having previously called PlanExplainsStop.  The way to finesse that is to have ShouldStop call DoPlanExplainsStop and then return IsPlanComplete.  The up side is no matter how you get to your ShouldStop, you only do the work to check what to do once, and all in one place.

(*) It might be more regular to redo the "PlanExplainsStop" every time we pop a plan to figure out the next plan we should ask ShouldStop of.  However, that would require other changes, for instance plans that only use other plans would have to start responding themselves to ExplainsStop, which they don't at present.